devito/finite_differences/differentiable.py

```diff
@@ -469,6 +469,64 @@ class Mod(DifferentiableOp, sympy.Mod):
     __sympy_class__ = sympy.Mod
 
 
+class IndexSum(DifferentiableOp):
+
+    """
+    A tensor contraction over one or more Dimensions.
+    """
+
+    is_commutative = True
+
+    def __new__(cls, expr, dimensions):
+        dimensions = as_tuple(dimensions)
+        if not dimensions:
+            return expr
+        for d in dimensions:
+            try:
+                if d.is_Dimension and is_integer(d.symbolic_size):
+                    continue
+            except AttributeError:
+                pass
+            raise ValueError("Expected Dimension with numeric size, "
+                             "got `%s` instead" % d)
+        for d in dimensions:
+            if d not in expr.free_symbols:
+                raise ValueError("Dimension `%s` must appear in `expr`" % d)
+        for i in expr.find(IndexSum):
+            for d in dimensions:
+                if d in i.dimensions:
+                    raise ValueError("Dimension `%s` already appears in a "
+                                     "nested tensor contraction" % d)
+
+        obj = sympy.Expr.__new__(cls, expr)
+        obj.dimensions = dimensions
+
+        return obj
+
+
+class Dot(IndexSum):
+
+    """
+    Dot-product, as the IndexSum of a binary Mul.
```

Comment:
Would probably say instead that this is a symbolic dot product of two symbolic vectors and add a little example